events/nats_message.go

```diff
@@ -0,0 +1,217 @@
+	}
+
+	m.sourceMetadata = metadata
+
+	return *m.sourceMetadata
+}
+
+// ID returns the nats message sequence number for the consumer.
+func (m *NatsMessage[T]) ID() string {
+	return strconv.FormatUint(m.metadata().Sequence.Consumer, base10)
+}
+
+// Topic returns the nats subject.
+func (m *NatsMessage[T]) Topic() string {
+	return m.source.Subject
+}
+
+// Message returns the decoded message object.
+func (m *NatsMessage[T]) Message() T {
+	return m.message
+}
+
+// Ack acks the message.
+func (m *NatsMessage[T]) Ack() error {
+	return m.source.Ack()
+}
+
+// Nak calls a Nak with the provided delay.
+func (m *NatsMessage[T]) Nak(delay time.Duration) error {
+	return m.source.NakWithDelay(delay)
+}
+
+// Term terminates the message from being processed again.
+func (m *NatsMessage[T]) Term() error {
+	return m.source.Term()
+}
+
+// Timestamp returns the timestamp of the message.
+func (m *NatsMessage[T]) Timestamp() time.Time {
+	return m.metadata().Timestamp
+}
+
+// Deliveries returns the number of times the message was delivered.
+func (m *NatsMessage[T]) Deliveries() uint64 {
+	return m.metadata().NumDelivered
+}
+
+// Error returns any error with the message.
+func (m *NatsMessage[T]) Error() error {
+	if m.err != nil {
+		return m.err
+	}
+
+	return nil
+}
+
+// Source returns the underlying nats message.
+func (m *NatsMessage[T]) Source() any {
+	return m.source
+}
+
+func (m *NatsMessage[T]) publish() error {
+	return m.conn.conn.PublishMsg(m.source)
+}
+
+func (m *NatsMessage[T]) request(ctx context.Context) (Message[AuthRelationshipResponse], error) {
+	if m.source.Reply == "" {
+		m.source.Reply = m.conn.conn.NewRespInbox()
+	}
+
+	nMsg, err := m.conn.conn.RequestMsgWithContext(ctx, m.source)
+	if err != nil {
+		// ensure we wrap no responder errors with ErrRequestNoResponders.
+		if errors.Is(err, nats.ErrNoResponders) {
+			return nil, fmt.Errorf("%w: %w", ErrRequestNoResponders, err)
+		}
+
+		return nil, err
+	}
+
+	respMsg := natsDecodeMessage[AuthRelationshipResponse](m.conn, nMsg)
+
+	return respMsg, nil
+}
+
+var _ Request[AuthRelationshipRequest, AuthRelationshipResponse] = (*NatsAuthRelationshipRequest)(nil)
+
+// NatsAuthRelationshipRequest implements Request for AuthRelationshipRequest / AuthRelationshipResponse
+type NatsAuthRelationshipRequest struct {
+	*NatsMessage[AuthRelationshipRequest]
+}
+
+// Reply responds to an AuthRelationshipRequest with an AuthRelationshipResponse.
+func (r *NatsAuthRelationshipRequest) Reply(ctx context.Context, message AuthRelationshipResponse) (Message[AuthRelationshipResponse], error) {
+	ctx, span := r.conn.tracer.Start(ctx, "events.Reply")
+
+	defer span.End()
+
+	if r.source.Reply == "" {
+		span.RecordError(ErrNatsMessageNoReplySubject)
+		span.SetStatus(codes.Error, ErrNatsMessageNoReplySubject.Error())
+
+		return nil, ErrNatsMessageNoReplySubject
+	}
+
+	if err := message.Validate(); err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
```

Comment:
We record errors here which is good, but don't do so consistently. Not a blocker here, but food for thought.

Reply:
fixed!